### PR TITLE
[compiler] Remove transitive memo check in validatePreserveMemo

### PR DESCRIPTION
**Pull request by mofeiZ**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30603
* __->__ #30630
* #30629
* #30602
* #30601

---
The original version of validatePreserveMemo tried to check that all manual memo declarations are *transitively* memoized. This may have been intended to stabilize `useEffect` by ensuring that optimized components trigger effects minimally.

This transitive memoization check only rarely affects compilation output. (Syncing this PR resulted in 12 changes out of tens of thousands of files). This is likely because many `StartMemoize` and `FinishMemoize` deps / decls are identifiers with no attached scopes (see #30603 for more complete explanation).

Meta internal reviewers: see [changes](https://fburl.com/phabricator/yzlich3a) from this PR

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/preserve-memo-validation/useCallback-dep-scope-pruned.expect.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  useIdentity(2);
+  mutate(x);
+  let t0;
+  if ($[0] !== x) {
+    t0 = () => [x, []];
+    $[0] = x;
+    $[1] = t0;
+  } else {
+    t0 = $[1];
+  }
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/preserve-memo-validation/useCallback-dep-scope-pruned.expect.md:53]: Note that this output correctly preserves manual memoization

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/preserve-memo-validation/useCallback-extended-contextvar-scope.expect.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  const t2 = x;
+  let t3;
+  if ($[1] !== a || $[2] !== t2) {
+    t3 = () => [a, x];
+    $[1] = a;
+    $[2] = t2;
+    $[3] = t3;
+  } else {
+    t3 = $[3];
+  }
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/preserve-memo-validation/useCallback-extended-contextvar-scope.expect.md:83]: This inferred memoization also preserves manual memo

File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidatePreservedManualMemoization.ts [hunk 3]
```diff
@@ -276,9 +276,16 @@ function validateInferredDep(
...
+   * Records all completed scopes (regardless of transitive memoization
+   * of scope dependencies)
+   *
+   * Both @scopes and @prunedScopes are live sets. We rely on iterating
+   * the reactive-ir in evaluation order, as they are used to determine
+   * whether scope dependencies / declarations have completed mutation.
+   */
   scopes: Set<ScopeId> = new Set();
   prunedScopes: Set<ScopeId> = new Set();
-  scopeMapping = new Map();
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/Validation/ValidatePreservedManualMemoization.ts:281]: (drive by dead code deletion)

**Review by mvitousek — APPROVED** [review]
nice

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/preserve-memo-validation/useCallback-dep-scope-pruned.expect.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+export const FIXTURE_ENTRYPOINT = {
+  fn: useFoo,
+  params: [3],
+};
+
+```
+      
+### Eval output
+(kind: exception) mutate is not defined
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/preserve-memo-validation/useCallback-dep-scope-pruned.expect.md:65]: :-)

**Review by josephsavona — COMMENTED** [review]
This seems reasonable. We should probably still fix up ValidateMemoizedEffectDeps and do a transitive check there, but I agree it's fine for ValidatePreserveManualMemo to be only as precise as the original memoization. Especially given the very low percentage of files impacted.

**Pull request merged**